compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/unpeel.mlir

```diff
@@ -0,0 +1,182 @@
+          memref.dealloc %alloc_3 : memref<2x2x1x1x4x4xf32, 2 : i32>
+          memref.dealloc %alloc_2 : memref<2x1x4x32xbf16, 1 : i32>
+          memref.dealloc %alloc_1 : memref<1x2x32x4xbf16, 1 : i32>
+          memref.dealloc %alloc_0 : memref<1x1x4x1x4x8xbf16, 2 : i32>
+          memref.dealloc %alloc : memref<1x1x1x4x8x4xbf16, 2 : i32>
+          return
+        }
+      }
+    }
+  }
+  util.func public @matmul(%arg0: !hal.buffer_view, %arg1: !hal.buffer_view) -> !hal.buffer_view attributes {iree.abi.stub, iree.reflection = {iree.abi.declaration = "sync func @matmul(%input0: tensor<8x128xbf16>, %input1: tensor<128x8xbf16>) -> (%output0: tensor<8x8xf32>)"}} {
+    %c256 = arith.constant 256 : index
+    %c2048 = arith.constant 2048 : index
+    %c0 = arith.constant 0 : index
+    %c128 = arith.constant 128 : index
+    %c8 = arith.constant 8 : index
+    %element_type_bf16 = hal.element_type<bf16> : i32
+    %dense_row_major = hal.encoding_type<dense_row_major> : i32
+    hal.buffer_view.assert<%arg0 : !hal.buffer_view> message("input0") shape([%c8, %c128]) type(%element_type_bf16) encoding(%dense_row_major)
+    %0 = stream.tensor.import on(#hal.device.affinity<@__device_0>) %arg0 : !hal.buffer_view -> tensor<8x128xbf16> in !stream.resource<external>{%c2048}
+    hal.buffer_view.assert<%arg1 : !hal.buffer_view> message("input1") shape([%c128, %c8]) type(%element_type_bf16) encoding(%dense_row_major)
+    %1 = stream.tensor.import on(#hal.device.affinity<@__device_0>) %arg1 : !hal.buffer_view -> tensor<128x8xbf16> in !stream.resource<external>{%c2048}
+    %result, %result_timepoint = stream.resource.alloca uninitialized on(#hal.device.affinity<@__device_0>) : !stream.resource<external>{%c256} => !stream.timepoint
+    %2 = stream.cmd.execute on(#hal.device.affinity<@__device_0>) await(%result_timepoint) => with(%0 as %arg2: !stream.resource<external>{%c2048}, %1 as %arg3: !stream.resource<external>{%c2048}, %result as %arg4: !stream.resource<external>{%c256}) {
+      stream.cmd.dispatch @matmul_dispatch_0::@amdaie_pdi_fb::@matmul_dispatch_0_matmul_8x8x128_bf16xbf16xf32 {
+        ro %arg2[%c0 for %c2048] : !stream.resource<external>{%c2048},
+        ro %arg3[%c0 for %c2048] : !stream.resource<external>{%c2048},
+        wo %arg4[%c0 for %c256] : !stream.resource<external>{%c256}
+      }
+    } => !stream.timepoint
+    %3 = stream.timepoint.await %2 => %result : !stream.resource<external>{%c256}
+    %4 = stream.tensor.export on(#hal.device.affinity<@__device_0>) %3 : tensor<8x8xf32> in !stream.resource<external>{%c256} -> !hal.buffer_view
+    util.return %4 : !hal.buffer_view
+  }
+}
+
+// Verify that the fill is written the peeled matmul is reabsoted, but the fill remains.
+// CHECK:         %[[FILL:.*]] = linalg.fill
+// CHECK:         scf.forall.in_parallel {
+// CHECK:           tensor.parallel_insert_slice %[[FILL]]
+// CHECK:         }
+// CHECK:         scf.for
+// CHECK-SAME:    %c0 to %c3 step %c1
```

Comment:
This is not clear to me what the output is supposed to look like. Could you expand the checks or attach the IR after this pass?

Reply:
Added links to before and after in the comment

Reply:
>  Thanks for sharing the IR! So far the IR seems good to me. It's good to further check the IR before LowerToAIE to see if anything suspicious. 

I studied the IR for a few hours yesterday, but didn't see anything suspicious. I'd appreciate it if you or @jtuyls could take a look at some point, I'm still not fully up-to-speed with the semantics for all the dma/object-fifo/connection ops, so I could be missing something obvious. But yeah I agree -- the IR change done in this PR's unpeeling loop looks fine/harmless. 

> What kind of numerical issues you saw? All zeros / all results are non-zeros but wrong?

Yup, all zeros. I tried a few things to diagnose further: I tried filling with `1` instead of `0`,  because I thought maybe the linalg.fill was somehow happening after the matmul, but that wasn't the case -- filling with 1 instead of 0 still resulted in the output being all 0. Which suggests no data is being written back to L3, maybe? I also tried removing the prologue entirely, but still get all 0 too. 

> About the motivation, I didn't quite get why this would help reduce program memory usage?

It's basically the same reason as outlining reduces PM. The outlining that @Abhishek-Varma  implemented creates a function which does the full 32x32 matmul. Take a look at the IR generated with function outlining and you'll see something like:

```
...
%24 = amdaie.core(%tile, in : [%21, %22], out : []) {
        func.call @generic_matmul_0_outlined(%alloc_0, %alloc, %alloc_3) : 
           (memref<1x1x4x8x4x8xbf16, 2 : i32>, 
            memref<1x1x8x4x8x4xbf16, 2 : i32>, 
            memref<1x1x8x8x4x4xf32, 2 : i32>) -> ()
        amdaie.end
}
...
```

i.e. instead of the ` linalg.generic` there is a `func.call`. Currently, you will see 3 such function calls in the IR: 1 for the prologue, 1 for the main body (call inside the sc.for loop) and 1 for the epilogue. 

This results in PM reduction, because PM will only contains the instructions for doing 1 32x32 matmul. These are alot of instructions saved, because llvm/peano unrolls the 32x32 matmul into lots of 4x4 matmuls. Without function outlining, these unrolled instructions will be duplicated 3 times in PM: once for the prologue, once for main, and once for epilogue. 

With the approach in this PR, I'm completely eliminating the matmul from the prologue and epilogue. So that even before Abhishek's function outlining, there is only 1 32x32 matmul in PM. 

Note that the scf.for loop has a trip count which is 2 larger than before with this PR. In theory if peano unrolled this outer scf.for loop, the approach in this PR wouldn't work. But it doesn't unroll this outer scf.for loop -- I've checked this, the number of appearances of `llvm.aie2.bf.mac16.conf` in the optimized code is exactly the same for K=2048 and K=4096 (if you run tests in run.py, the optimized llvm IR after unrolling is in `module_matmul_dispatch_0_amdaie_pdi_fb/input.opt.ll`). 

It would seem crazy to me if peano did unroll this outer scf.for loop, because it contains lock accesses and releases, so I don't think there would be any advantage to peano unrolling it (@jtuyls might have some idea here). 

Finally I should mention what the advantage of this approach over function outlining: there will be no performance hit caused by function calls.

Reply:
Okay, thanks for the explanation. In regarding of the all zeros, there could be different reasons behind that. Probably check the access patterns and locks in terms of racing condition? Could you please share the dump IR after all passes? I could take a look as well.

Reply:
>  Could you please share the dump IR after all passes? I could take a look as well.

Thanks! I've uploaded it https://github.com/user-attachments/files/17757030/with_unpeel_before_all.txt